nibabel/cmdline/roi.py

```diff
@@ -0,0 +1,88 @@
+import sys
+import os
+import argparse
+import nibabel as nb
+
+
+def lossless_slice(img, slicers):
+    if not nb.imageclasses.spatial_axes_first(img):
+        raise ValueError("Cannot slice an image that is not known to have spatial axes first")
+
+    scaling = hasattr(img.header, 'set_slope_inter')
+
+    data = img.dataobj._get_unscaled(slicers) if scaling else img.dataobj[slicers]
+    roi_img = img.__class__(data, affine=img.slicer.slice_affine(slicers), header=img.header)
+
+    if scaling:
+        roi_img.header.set_slope_inter(img.dataobj.slope, img.dataobj.inter)
+    return roi_img
+
+
+def parse_slice(crop, allow_step=True):
+    if crop is None:
+        return slice(None)
+    start, stop, *extra = [int(val) if val else None for val in crop.split(":")]
+    if len(extra) > 1:
+        raise ValueError(f"Cannot parse specification: {crop}")
+    if not allow_step and extra and extra[0] not in (1, None):
+        raise ValueError(f"Step entry not permitted: {crop}")
+
+    step = extra[0] if extra else None
+    if step not in (1, -1, None):
+        raise ValueError(f"Downsampling is not supported: {crop}")
+
+    return slice(start, stop, step)
+
+
+def sanitize(args):
+    # Argparse likes to treat "-1:..." as a flag
+    return [f' {arg}' if arg[0] == '-' and ":" in arg else arg
+            for arg in args]
+
+
+def main(args=None):
+    if args is None:
+        args = sys.argv[1:]
+    parser = argparse.ArgumentParser(
+        description="Crop images to a region of interest",
+        epilog="If a start or stop value is omitted, the start or end of the axis is assumed.")
+    parser.add_argument('--version', action='version', version=nb.__version__)
+    parser.add_argument("-i", metavar="I1:I2[:-1]",
+                        help="Start/stop [flip] along first axis (0-indexed)")
+    parser.add_argument("-j", metavar="J1:J2[:-1]",
+                        help="Start/stop [flip] along second axis (0-indexed)")
+    parser.add_argument("-k", metavar="K1:K2[:-1]",
+                        help="Start/stop [flip] along third axis (0-indexed)")
+    parser.add_argument("-t", metavar="T1:T2", help="Start/stop along fourth axis (0-indexed)")
+    parser.add_argument("in_file", help="Image file to crop")
+    parser.add_argument("out_file", help="Output file name")
+
```

Comment:
should this show the help if no arguments are provided?
```suggestion
    if not args:
        parser.print_help()
        sys.exit(1)
```

Reply:
You do at least get usage:

```
$ nib-roi
usage: nib-roi [-h] [--version] [-i I1:I2[:-1]] [-j J1:J2[:-1]]
               [-k K1:K2[:-1]] [-t T1:T2]
               in_file out_file
nib-roi: error: the following arguments are required: in_file, out_file
$ nib-roi --help
usage: nib-roi [-h] [--version] [-i I1:I2[:-1]] [-j J1:J2[:-1]]
               [-k K1:K2[:-1]] [-t T1:T2]
               in_file out_file

Crop images to a region of interest

positional arguments:
  in_file        Image file to crop
  out_file       Output file name

optional arguments:
  -h, --help     show this help message and exit
  --version      show program's version number and exit
  -i I1:I2[:-1]  Start/stop [flip] along first axis (0-indexed)
  -j J1:J2[:-1]  Start/stop [flip] along second axis (0-indexed)
  -k K1:K2[:-1]  Start/stop [flip] along third axis (0-indexed)
  -t T1:T2       Start/stop along fourth axis (0-indexed)

If a start or stop value is omitted, the start or end of the axis is assumed.
```

We don't have a terribly consistent style, to go by:

```
$ nib-ls; echo $?

0
$ nib-conform; echo $?
usage: nib-conform [-h] [--out-shape OUT_SHAPE OUT_SHAPE OUT_SHAPE]
                   [--voxel-size VOXEL_SIZE VOXEL_SIZE VOXEL_SIZE]
                   [--orientation ORIENTATION] [-f] [-V]
                   infile outfile
nib-conform: error: the following arguments are required: infile, outfile
2
$ nib-diff; echo $?
Traceback (most recent call last):
  File "/home/cjmarkie/anaconda3/envs/nibabel_dev/bin/nib-diff", line 11, in <module>
    load_entry_point('nibabel', 'console_scripts', 'nib-diff')()
  File "/home/cjmarkie/Projects/nipy/nibabel/nibabel/cmdline/diff.py", line 362, in main
    dtype=opts.dtype
  File "/home/cjmarkie/Projects/nipy/nibabel/nibabel/cmdline/diff.py", line 315, in diff
    assert len(files) >= 2, "Please enter at least two files"
AssertionError: Please enter at least two files
1
$ nib-tck2trk; echo $?
usage: nib-tck2trk [-h] [-f] anatomy tractogram [tractogram ...]
nib-tck2trk: error: the following arguments are required: anatomy, tractogram
2
```

But on the whole I think I'm okay with argparse defaults.